src/jupyter_contrib_nbextensions/nbextensions/exercise/main.js

```diff
@@ -34,169 +34,116 @@ define([
+        if (ctrl.length > 0) return ctrl;
+        var locked = cell.metadata.solution === 'hidden';
+        ctrl = $('<div class="exercise fa">')
+            .prependTo(cell.element)
+            .on('click', click_solution_lock);
+        element_set_locked(cell, locked);
+        return ctrl;
+    }
 
-    load_css('./main.css');
+    function remove_element(cell) {
+        cell.element.find('.exercise').remove();
+    }
 
-    // ***************** Keyboard shortcuts ******************************
-    var add_cmd_shortcuts = {
-        'Alt-S': {
-            help: 'Define Solution (Exercise)',
-            help_index: 'ht',
-            handler: function(event) {
-                hide_solutions();
-                return false;
-            }
-        }
+    function element_set_locked(cell, locked) {
+        return cell.element.find('.exercise')
+            .toggleClass('fa-plus-square-o', locked)
+            .toggleClass('fa-minus-square-o', !locked);
     }
-    IPython.keyboard_manager.command_shortcuts.add_shortcuts(add_cmd_shortcuts);
-    
-    /**
-     * Display existing solutions at startup
-     *
-     */
-    var cells = IPython.notebook.get_cells();
-    var found_solution = false;
-    for(var i in cells){
-        var cell = cells[i];
-        if (found_solution == true && typeof cell.metadata.solution != "undefined" && cell.metadata.solution_first !=true) {
-            cell.element.toggleClass('hidden', cell.metadata.solution  === 'hidden');
-        } else {
-            found_solution = false
-        }
 
-        if (found_solution == false && typeof cell.metadata.solution != "undefined") {
-            if (cell.metadata.solution=="hidden") var el = $('<div id="lock" class="fa fa-plus-square-o">');
-            else var el = $('<div id="lock" class="fa fa-minus-square-o">');
-            cell.element.prepend(el);
-            el.click( click_solution_lock);
-            found_solution = true;
-        }
+    function refresh_exercises() {
+        var in_exercise = false;
+        IPython.notebook.get_cells().forEach(function(cell) {
+            if (in_exercise && cell.metadata.solution !== undefined && !cell.metadata.solution_first) {
+                cell.element.toggleClass('hidden', cell.metadata.solution  === 'hidden');
+            } else {
+                in_exercise = false;
+            }
+            if (!in_exercise && cell.metadata.solution !== undefined) {
+                in_exercise = true;
+                add_element(cell);
+            }
+        });
     }
 
-}
+    function load_ipython_extension() {
+        // add css
+        $('<link rel="stylesheet" type="text/css">')
+            .attr('href', requirejs.toUrl('./main.css'))
+            .appendTo('head');
+
+        // Hide/display existing solutions at startup
+        events.on('notebook_loaded.Notebook', refresh_exercises);
+        if (IPython.notebook._fully_loaded) refresh_exercises();
```

Comment:
Isn't necessary to check that `IPython.notebook != undefined`?

Reply:
notebook is (at least in recent versions, maybe not in 3.x?) always _defined_ before any nbextensions are called, even if not fully loaded